# Exploring the data

Ideas:
- Train a model that eventually matches with real world rankings
- Determine the features that have the most impact based on the weights

In [2]:
import pandas as pd
import altair as alt

## Mapping labels with string classification

![](https://www.startpage.com/av/proxy-image?piurl=https%3A%2F%2Ftse3.explicit.bing.net%2Fth%3Fid%3DOIP.8wHctLr_Dq71EwdJ8ydV3gHaEK%26pid%3DApi&sp=1741834617Ta555c5c0c05aefccefb18821829f8a1fad2051ba2c2f2699f7b018685c719e62)

Un-doing the categorical encoding already presented to get a better understanding of the data.

In [3]:
idcAboutStates_df = pd.read_csv("Labels.csv")

## I LOVE CLEANING UP DATA

Some of the columns in `data_df` have extra spaces in random places. Causing `replace()` to fail for those columns.
So annoying

![](https://media1.tenor.com/m/z4Pjus3zEcsAAAAd/dad-coraline.gif)


In [4]:
# Filter out state abbreviations from our label data because it's irrelevant
idcAboutStates_df = (
    idcAboutStates_df[
        ~idcAboutStates_df["VariableName"]
        .str.contains("abbreviation", regex=True, case=False)]
        .astype({"Value": int, "ValueLabel": str}))

unique_variables = idcAboutStates_df["VariableName"].unique()

In [128]:
data_df = pd.read_csv("data.csv").drop(columns="UnitID")

In [129]:
for row in idcAboutStates_df.itertuples():
    try:
        data_df[row.VariableName].replace({row.Value : row.ValueLabel}, inplace=True)
    # Skip over labels that don't exist on the data
    except KeyError as e:
        print(f"wtf man: {e}")

/var/folders/7c/62z3f6vd7_qcckfpqg8f6wnr0000gn/T/ipykernel_89460/1565544624.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data_df[row.VariableName].replace({row.Value : row.ValueLabel}, inplace=True)


In [108]:
data_df

Institution Name  \
0                  Aaniiih Nakoda College   
1    Abraham Baldwin Agricultural College   
2                  Adams State University   
3                  Aims Community College   
4                Alabama A & M University   
..                                    ...   
751   Wright State University-Main Campus   
752                 Yakima Valley College   
753                       Yavapai College   
754           Youngstown State University   
755                    Zane State College   

       Institution (entity) name (HD2023)  \
0                  Aaniiih Nakoda College   
1    Abraham Baldwin Agricultural College   
2                  Adams State University   
3                  Aims Community College   
4                Alabama A & M University   
..                                    ...   
751   Wright State University-Main Campus   
752                 Yakima Valley College   
753                       Yavapai College   
754           Youngstown State University   
755                    Zane State College   

     Historically Black College or University (HD2023)  \
0                                                    2   
1                                                    2   
2                                                    2   
3                                                    2   
4                                                    1   
..                                                 ...   
751                                                  2   
752                                                  2   
753                                                  2   
754                                                  2   
755                                                  2   

     Tribal college (HD2023)  Carnegie Classification 2021: Basic (HD2023)  \
0                          1                                            33   
1                          2                                            23   
2                          2                                            18   
3                          2                                             3   
4                          2                                            18   
..                       ...                                           ...   
751                        2                                            16   
752                        2                                            14   
753                        2                                             2   
754                        2                                            18   
755                        2                                             6   

     Institution grants a medical degree (HD2023) State abbreviation (HD2023)  \
0                                               2                          MT   
1                                               2                          GA   
2                                               2                          CO   
3                                               2                          CO   
4                                               2                          AL   
..                                            ...                         ...   
751                                             1                          OH   
752                                             2                          WA   
753                                             2                          AZ   
754                                             2                          OH   
755                                             2                          OH   

     Carnegie Classification 2021: Undergraduate Profile (HD2023)  \
0                                                    3              
1                                                    6              
2                                                    7              
3                                                    1      

In [8]:
count_carnegie = data_df['Carnegie Classification 2021: Undergraduate Profile (HD2023)'].value_counts().reset_index()
count_carnegie.rename(columns={count_carnegie.columns[0] : "category"}, inplace=True)

### Carnegie Classification

Majority of the institutions in the list are 4 year institutions, with 4 year, full time and with a higher transfer rate being the most common

In [9]:
alt.Chart(count_carnegie.nlargest(5, 'count')).mark_arc().encode(
theta="count",
color=alt.Color("category", legend=alt.Legend(labelLimit=0, labelFontSize=15)),
tooltip=[
        alt.Tooltip("category", title="Carnegie Classification"),
        alt.Tooltip("count", title="Count")
    ]
).properties(title="Carnegie Classification")

alt.Chart(...)

In [38]:
data_df.nlargest(40, "U.S. Nonresident total (EF2023A  All students total)")

UnitID                                   Institution Name  \
174  139755        Georgia Institute of Technology-Main Campus   
597  145637            University of Illinois Urbana-Champaign   
13   104151          Arizona State University Campus Immersion   
396  243780                      Purdue University-Main Campus   
620  170976                   University of Michigan-Ann Arbor   
653  227216                          University of North Texas   
682  236948            University of Washington-Seattle Campus   
565  110680                 University of California-San Diego   
559  110635                  University of California-Berkeley   
360  214777          Pennsylvania State University-Main Campus   
690  240444                    University of Wisconsin-Madison   
525  228787                  The University of Texas at Dallas   
545  196088                              University at Buffalo   
562  110662               University of California-Los Angeles   
560  110644                     University of California-Davis   
335  204796                  Ohio State University-Main Campus   
523  228769               The University of Texas at Arlington   
561  110653                    University of California-Irvine   
671  137351                        University of South Florida   
502  228723             Texas A & M University-College Station   
204  151351                     Indiana University-Bloomington   
627  174066                University of Minnesota-Twin Cities   
524  228778                  The University of Texas at Austin   
408  186380                   Rutgers University-New Brunswick   
584  134130                              University of Florida   
574  201885               University of Cincinnati-Main Campus   
154  133951                   Florida International University   
590  225511                              University of Houston   
613  163286                University of Maryland-College Park   
267  171100                          Michigan State University   
615  166629                University of Massachusetts-Amherst   
170  232186                            George Mason University   
197  225423                          Houston Community College   
595  145600                     University of Illinois Chicago   
711  233921  Virginia Polytechnic Institute and State Unive...   
570  176965                     University of Central Missouri   
310  199193         North Carolina State University at Raleigh   
552  104179                              University of Arizona   
476  196097                             Stony Brook University   
578  129020                          University of Connecticut   

                    Institution (entity) name (HD2023)  \
174        Georgia Institute of Technology-Main Campus   
597            University of Illinois Urbana-Champaign   
13           Arizona State University Campus Immersion   
396                      Purdue University-Main Campus   
620                   University of Michigan-Ann Arbor   
653                          University of North Texas   
682            University of Washington-Seattle Campus   
565                 University of California-San Diego   
559                  University of California-Berkeley   
360          Pennsylvania State University-Main Campus   
690                    University of Wisconsin-Madison   
525                  The University of Texas at Dallas   
545                              University at Buffalo   
562               University of California-Los Angeles   
560                     University of California-Davis   
335                  Ohio State University-Main Campus   
523               The University of Texas at Arlington   
561                    University of California-Irvine   
671                        University of South Florida   
502             Texas A & M University-College Station   
204                     Indiana University-Bloomington   
627                Uni

## Column Groupings

To avoid the possibility of carpal tunnel, we group the columns.
To get more information about the groupings, see the Notes notebook

Grand total (EF2023  All students total)  \
0                                         133   
1                                        3768   
2                                        2950   
3                                        7529   
4                                        6614   
..                                        ...   
751                                      9884   
752                                      3523   
753                                      7200   
754                                     11040   
755                                      1619   

     Grand total (EF2023  All students  Undergraduate total)  \
0                                                  133         
1                                                 3768         
2                                                 1576         
3                                                 7529         
4                                                 5845         
..                                                 ...         
751                                               7012         
752                                               3523         
753                                               7200         
754                                               8436         
755                                               1619         

     Grand total (EF2023  All students  Graduate and First professional)  
0                                                  NaN                    
1                                                  NaN                    
2                                               1374.0                    
3                                                  NaN                    
4                                                769.0                    
..                                                 ...                    
751                                             2872.0                    
752                                                NaN                    
753                                                NaN                    
754                                             2604.0                    
755                                                NaN                    

[756 rows x 3 columns]

## Add real world rankings

Add columns on where an institution ranks

Each real world ranking such as:
- U.S. News Best Colleges
- Wall Street Journal
- Princeton Review
- Forbes
- Washington Monthly

Analyze and look for patterns for each rankings with the variables from the original data

### External College Rankings

External rankings use different annotations so an exact merge is not always possible.
Use fuzzy matching with a threshold of 85 for accuracy

#### Niche

https://www.niche.com/api/renaissance/results/?type=private&type=public&listURL=best-colleges&page=1&searchType=college&limit=800

TU ranks 341. 

I'm cooked ![](https://cdn.betterttv.net/emote/59f27b3f4ebd8047f54dee29/3x.webp)

#### Forbes 2025

TU ranks 174

#### Times Higher Ed

TU ranks 391

In [49]:
from json import load
with open("Ranking_datasets/niche-800.json", "r") as f:
    niche_json = load(f)

with open("Ranking_datasets/Forbes-Ranking-2025.json") as f:
    forbes_json = load(f)

with open("Ranking_datasets/timeshighered-2022.json") as f:
    times_json = load(f)

niche_rankings = {}
forbes_rankings = {}
highered_rankings = {}
for idx, university in enumerate(niche_json["entities"]):
    niche_rankings[university["content"]["entity"]["name"]] = idx + 1

niche_df = pd.DataFrame(niche_rankings.items(), columns=["Institution", "Niche_Ranking"])

for university in forbes_json["organizationList"]["organizationsLists"]:
    forbes_rankings[university["organizationName"]] = university["rank"]

forbes_df = pd.DataFrame(forbes_rankings.items(), columns=["Institution", "Forbes_Ranking"])

for university in times_json["data"]:
    highered_rankings[university["name"]] = university["rank_order"]

highered_df = pd.DataFrame(highered_rankings.items(), columns=["Institution", "HigherEd_Ranking"])

In [130]:
from thefuzz import fuzz

# ngl, I had to use Claude because laziness and it's pretty common anyway
def simple_fuzzy_merge(df_left, df_right, left_on="Institution Name", right_on = "Institution", threshold=97):
    """
    Performs a simplified fuzzy merge between two dataframes using fuzzywuzzy.
    
    Parameters:
    -----------
    df_left : pandas DataFrame
        Left dataframe to merge
    df_right : pandas DataFrame
        Right dataframe to merge
    left_on : str
        Column name from left dataframe to match on
    right_on : str
        Column name from right dataframe to match on
    threshold : int, default 90
        Minimum similarity score to consider a match (0-100)
        
    Returns:
    --------
    pandas DataFrame
        Merged dataframe with additional column showing match score
    """
    # Create a copy of dataframes to avoid modifying originals
    left_df = df_left.copy().reset_index().rename(columns={'index': 'left_idx'})
    right_df = df_right.copy().reset_index().rename(columns={'index': 'right_idx'})
    
    # Create an empty list to store matches
    matches = []
    
    # For each value in the left dataframe
    for _, left_row in left_df.iterrows():
        left_value = str(left_row[left_on]).lower()
        left_idx = left_row['left_idx']
        
        # Calculate similarity scores with all values in the right dataframe
        right_df['score'] = right_df[right_on].apply(
            lambda x: fuzz.token_sort_ratio(left_value, str(x).lower())
        )
        
        # Get the best match if it's above the threshold
        best_match = right_df[right_df['score'] >= threshold].sort_values('score', ascending=False).head(1)
        
        if not best_match.empty:
            matches.append({
                'left_idx': left_idx,
                'right_idx': best_match['right_idx'].values[0],
                'score': best_match['score'].values[0]
            })
    
    # If no matches were found, return empty dataframe
    if not matches:
        return pd.DataFrame()
    
    # Create a dataframe from matches
    matches_df = pd.DataFrame(matches)
    
    # Merge the original dataframes based on the indices from matches_df
    result = pd.merge(
        df_left.loc[matches_df['left_idx']].reset_index(drop=True),
        df_right.loc[matches_df['right_idx']].reset_index(drop=True),
        left_index=True, 
        right_index=True,
        suffixes=('_left', '_right')
    )
    
    # Add the match score
    result['fuzzy_score'] = matches_df['score'].values
    
    return result

merged_niche = simple_fuzzy_merge(data_df, niche_df)
merged_forbes = simple_fuzzy_merge(data_df, forbes_df)
merged_highered = simple_fuzzy_merge(data_df, highered_df)

In [46]:
niche_df

Institution  Niche_Ranking  fuzz_score
0      Massachusetts Institute of Technology              1          79
1                            Yale University              2          81
2                        Stanford University              3          86
3                         Harvard University              4          82
4                          Dartmouth College              5          71
..                                       ...            ...         ...
795     Commonwealth University - Bloomsburg            796          69
796                    High Point University            797          76
797  West Virginia University at Parkersburg            798         100
798             Austin Peay State University            799         100
799                       Norwich University            800          76

[800 rows x 3 columns]

#### How TU compares to other Maryland colleges

In [118]:

merged_forbes.columns[88:97]

Index(['Asian total (EF2023A  All students total)',
       'Black or African American total (EF2023A  All students total)',
       'Hispanic total (EF2023A  All students total)',
       'Native Hawaiian or Other Pacific Islander total (EF2023A  All students total)',
       'White total (EF2023A  All students total)',
       'Two or more races total (EF2023A  All students total)',
       'Race/ethnicity unknown total (EF2023A  All students total)',
       'U.S. Nonresident total (EF2023A  All students total)',
       'All students enrolled (EF2023A_DIST  Undergraduate total)'],
      dtype='object')

In [ ]:
column_ranges = {
    "Degrees Conferred": (16, 21),
    "Financial Aid": (21, 27),
    "Student Success": (28, 32),
    "School Finance": (33, 63),
    "Library": (64, 71),
    "Admissions": (72, 74),
    "Race": (87, 96),
    "Population": (81, 84),
}

# Insert Institution Name for each dictionary values
column_groups = {
    name: data_df.columns[start:end].insert(0, data_df.columns[0])
    for name, (start, end) in column_ranges.items()
}

column_ranges.items()


dict_items([('Degrees Conferred', (16, 21)), ('Financial Aid', (21, 27)), ('Student Success', (28, 32)), ('School Finance', (33, 63)), ('Library', (64, 71)), ('Admissions', (72, 74)), ('Race', (87, 96)), ('Population', (81, 84))])

In [132]:
merged_forbes[merged_forbes["State abbreviation (HD2023)"] == "MD"].sort_values(by="Forbes_Ranking")[column_groups["Degrees Conferred"]]

Institution Name  \
134      University of Maryland-College Park   
99                         Towson University   
133  University of Maryland-Baltimore County   
79                      Salisbury University   
87            St. Mary's College of Maryland   
132     University of Maryland Global Campus   
107                  University of Baltimore   

     Grand total (C2023_A  First major  Grand total  Bachelor's degree)  \
134                                             8075.0                    
99                                              4064.0                    
133                                             2419.0                    
79                                              1605.0                    
87                                               291.0                    
132                                             7843.0                    
107                                              373.0                    

     Grand total (C2023_A  First major  Grand total  Master's degree)  \
134                                             3078.0                  
99                                               844.0                  
133                                             1000.0                  
79                                               287.0                  
87                                                14.0                  
132                                             3582.0                  
107                                              341.0                  

     Grand total (C2023_A  First major  Grand total  Doctor's degree - research/scholarship )  \
134                                              631.0                                          
99                                                12.0                                          
133                                              100.0                                          
79                                                 5.0                                          
87                                                 NaN                                          
132                                                NaN                                          
107                                                7.0                                          

     Grand total (C2023_A  First major  Grand total  Doctor's degree - professional practice )  \
134                                               39.0                                           
99                                                38.0                                           
133                                                NaN                                           
79                                                 8.0                                           
87                                                 NaN                                           
132                                               41.0                                           
107                                              196.0                                           

     Grand total (C2023_A  First major  Grand total  Doctor's degree - other )  
134                                                NaN                          
99                                                 NaN                          
133                                                NaN                          
79                                                 NaN                          
87                                                 NaN                          
132                                                NaN                          
107                                                NaN

## Data Preprocessing

In [ ]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

target_col = [col for col in merged_forbes.select_dtypes(include=["number"]).columns if col not in ["Forbes_Ranking", "fuzzy_score"]]
merged_forbes[target_col] = scaler.fit_transform(merged_forbes[target_col])

In [138]:
merged_forbes

Institution Name  \
0                         Appalachian State University   
1                                    Auburn University   
2                               Boise State University   
3    California Polytechnic State University-San Lu...   
4       California State Polytechnic University-Pomona   
..                                                 ...   
176                           West Virginia University   
177                      Western Washington University   
178                           Wichita State University   
179                                     William & Mary   
180          William Paterson University of New Jersey   

                    Institution (entity) name (HD2023)  \
0                         Appalachian State University   
1                                    Auburn University   
2                               Boise State University   
3    California Polytechnic State University-San Lu...   
4       California State Polytechnic University-Pomona   
..                                                 ...   
176                           West Virginia University   
177                      Western Washington University   
178                           Wichita State University   
179                                     William & Mary   
180          William Paterson University of New Jersey   

    Historically Black College or University (HD2023) Tribal college (HD2023)  \
0                                                  No                      No   
1                                                  No                      No   
2                                                  No                      No   
3                                                  No                      No   
4                                                  No                      No   
..                                                ...                     ...   
176                                                No                      No   
177                                                No                      No   
178                                                No                      No   
179                                                No                      No   
180                                                No                      No   

          Carnegie Classification 2021: Basic (HD2023)  \
0    Master's Colleges & Universities: Larger Programs   
1    Doctoral Universities: Very High Research Acti...   
2      Doctoral Universities: High Research Activity\n   
3    Master's Colleges & Universities: Larger Programs   
4    Master's Colleges & Universities: Larger Programs   
..                                                 ...   
176  Doctoral Universities: Very High Research Acti...   
177  Master's Colleges & Universities: Larger Programs   
178    Doctoral Universities: High Research Activity\n   
179    Doctoral Universities: High Research Activity\n   
180  Master's Colleges & Universities: Larger Programs   

    Institution grants a medical degree (HD2023) State abbreviation (HD2023)  \
0                                             No                          NC   
1                                            Yes                          AL   
2                                             No                          ID   
3                                             No                          CA   
4                                             No                          CA   
..                                           ...                         ...   
176                                          Yes                          WV   
177                                           No                          WA   
178                                           No                          KS   
179                                           No                          VA   
180                                           No                          NJ   

    Carnegie 